Sources/XcodeGenKit/CarthageDependencyResolver.swift

```diff
@@ -0,0 +1,165 @@
+                let nonExistentDependencies = target.dependencies.filter(nonExistingDepsPredicate)
+                for dependency in nonExistentDependencies {
+                    switch dependency.type {
+                    case .carthage:
+                        if dependency.includeRelated == true {
+                            relatedDependencies(for: dependency, in: target.platform)
+                                .filter(nonExistingDepsPredicate)
+                                .forEach { frameworks[$0.reference] = $0 }
+                        } else {
+                            frameworks[dependency.reference] = dependency
+                        }
+                    case .target:
+                        if let projectTarget = project.getProjectTarget(dependency.reference) {
+                            if let dependencyTarget = projectTarget as? Target {
+                                if topLevelTarget.platform == dependencyTarget.platform {
+                                    queue.append(projectTarget)
+                                }
+                            } else {
+                                queue.append(projectTarget)
+                            }
+                        }
+                    default:
+                        break
+                    }
+                }
+            } else if let aggregateTarget = projectTarget as? AggregateTarget {
+                for dependencyName in aggregateTarget.targets {
+                    if let projectTarget = project.getProjectTarget(dependencyName) {
+                        queue.append(projectTarget)
+                    }
+                }
+            }
+
+            visitedTargets.update(with: projectTarget.name)
+        }
+
+        return frameworks.sorted(by: { $0.key < $1.key }).map { $0.value }
+    }
+    
+    /// Reads the .version file generated for a given Carthage dependency
+    /// and returns a list of its related dependencies including self
+    func relatedDependencies(for dependency: Dependency, in platform: Platform) -> [Dependency] {
+        guard dependency.type == .carthage else { return [] }
+        let references: [CarthageVersionFile.Reference] = {
+            if let cachedVersionFile = carthageCachedRelatedDependencies[dependency.reference] {
+                return cachedVersionFile.references(for: platform)
+            }
+            let buildPath = project.basePath + "\(self.baseBuildPath)/.\(dependency.reference).version"
+            guard
+                let data = try? buildPath.read(),
+                let carthageVersionFile = try? JSONDecoder().decode(CarthageVersionFile.self, from: data) else {
+                    // No .version file or we've been unable to parse
+                    // so fail gracefully by returning an empty array
+                    return []
+            }
+            return carthageVersionFile.references(for: platform)
```

Comment:
What if there is no version file or it fails parsing? Then we still want to return the main dependency

Reply:
Nice catch, I'll work on it.